lgc/include/lgc/state/PalMetadata.h

```diff
@@ -147,17 +147,38 @@ class PalMetadata {
   // Erase the color export info
   void eraseColorExportInfo();
 
-  // Finalize PAL metadata for pipeline.
-  // TODO Shader compilation: The idea is that this will be called at the end of a pipeline compilation, or in
-  // an ELF link, but not at the end of a shader/half-pipeline compile.
-  void finalizePipeline();
+  // Finalize PAL metadata for pipeline, part-pipeline or shader compilation.
+  void finalizePipeline(bool isWholePipeline);
 
   // Updates the PS register information that depends on the exports.
   void updateSpiShaderColFormat(llvm::ArrayRef<ColorExportInfo> exps, bool hasDepthExpFmtZero, bool killEnabled);
 
   // Sets the finalized 128-bit cache hash.  The version identifies the version of LLPC used to generate the hash.
   void setFinalized128BitCacheHash(const lgc::Hash128 &finalizedCacheHash, const llvm::VersionTuple &version);
 
+  // Store the fragment shader input mapping information for a fragment shader being compiled by itself (partial
+  // pipeline compilation).
+  void addFragmentInputInfo(llvm::ArrayRef<unsigned> data1, llvm::ArrayRef<unsigned> data2);
+
+  // Check whether we have FS input mappings, and thus whether we're doing part-pipeline compilation of the
+  // pre-FS part of the pipeline.
+  bool haveFsInputMappings();
+
+  // In part-pipeline compilation, get a blob of data representing the FS input mappings that can be used by the
+  // client in a hash. The resulting data is owned by the PalMetadata object, and lives until the PalMetadata
+  // object is destroyed or another call is made to getFsInputMappings.
+  llvm::StringRef getFsInputMappings();
+
+  // In part-pipeline compilation, retrieve the FS input mappings into the provided vectors, and delete them
+  // from the PAL metadata so they do not appear in the final ELF.
+  void retrieveFragmentInputInfo(llvm::SmallVectorImpl<unsigned> &data1, llvm::SmallVectorImpl<unsigned> &data2);
+
+  // In part-pipeline compilation, copy any metadata needed from the "other" pipeline's PAL metadata into ours.
+  void setOtherPartPipeline(PalMetadata *other);
```

Comment:
nit: Can `other` be null? If not, consider using a reference instead.

Reply:
LLVM does seem to be heading in the direction of using a reference rather than a pointer when you know it can't be null. But LLPC/LGC code doesn't do that very much. It felt to me that it would be better to leave it as a pointer given the context of where this is used.

Reply:
I think that there's a split within LLVM on this: IR constructs (Values, Blocks, Functions, Types, etc) are passed by pointers, but I think that most other analysis objects have been used with references for a very long time (e.g., domtree, analysis manager). I think this makes sense for IR stuff as they have their own distinct memory management and most relevant APIs operate on that (e.g., casts that return nullptr when unsuccessful), but I don't see benefits of using pointers with types/objects that don't  benefit from that.